lxd/iptables/xtables.go

```diff
@@ -0,0 +1,337 @@
+package iptables
+
+import (
+	"encoding/hex"
+	"fmt"
+	"net"
+	"strings"
+
+	deviceConfig "github.com/lxc/lxd/lxd/device/config"
+	"github.com/lxc/lxd/lxd/util"
+	"github.com/lxc/lxd/shared"
+)
+
+// XTables is an implmentation of LXD firewall using {ip, ip6, eb}tables
+type XTables struct{}
+
+// Lower-level Functions
+
+// NetworkClear removes network rules.
+func (xt XTables) NetworkClear(protocol string, comment string, table string) error {
+	return NetworkClear(protocol, comment, table)
+}
+
+// InstanceClear removes rules all rules for the given instance.
+func (xt XTables) InstanceClear(protocol string, comment string, table string) error {
+	return ContainerClear(protocol, comment, table)
+}
+
+// VerifyIPv6Module checks to see if the ipv6 kernel module is present.
+func (xt XTables) VerifyIPv6Module() error {
+	// Check br_netfilter is loaded and enabled for IPv6.
+	sysctlPath := "net/bridge/bridge-nf-call-ip6tables"
+	sysctlVal, err := util.SysctlGet(sysctlPath)
+	if err != nil {
+		return fmt.Errorf("Error reading net sysctl %s: %v", sysctlPath, err)
+	}
+
+	if sysctlVal != "1\n" {
+		return fmt.Errorf("security.ipv6_filtering requires br_netfilter and sysctl net.bridge.bridge-nf-call-ip6tables=1")
+	}
+
+	return nil
+}
+
+// Proxy Functions
+
+// ProxySetupNAT creates a default NAT setup.
+func (xt XTables) ProxySetupNAT(protocol string, iptablesComment string, connType, address, port string, ipAddr net.IP, cPort string) error {
```

Comment:
`protocol` should instead be `family` and other functions making similar use be updated.
`protocol` should refer to the IP protocol (udp or tcp) instead.

Reply:
And I'd rename `ipAddr` to `destAddr` and `cPort` to `destPort`, would make things a bit clearer I think.

Reply:
`iptablesComment` should probably be renamed to `comment` for consistency and also be moved to being the last argument.

Reply:
Function should be renamed to `InstanceDeviceProxyNAT`.